File: src/tts/edge.mjs
import { spawn } from 'node:child_process';
import path from 'node:path';
import { ROOT } from '../config.mjs';
import { logger } from '../log.mjs';

const log = logger('edge-tts');
const HELPER = path.join(ROOT, 'src', 'tts', 'edge.py');

/**
 * Indonesian voices available without an account.
 *
 * Both are Azure neural voices trained on Indonesian, not English models
 * reading Indonesian phonemes, which is what makes pepet and taling land where
 * a native speaker expects them.
 */
export const VOICES = [
  { id: 'id-ID-GadisNeural', label: 'Gadis — perempuan, hangat', gender: 'perempuan' },
  { id: 'id-ID-ArdiNeural', label: 'Ardi — laki-laki, tenang', gender: 'laki-laki' },
];

export const DEFAULT_VOICE = 'id-ID-GadisNeural';

/**
 * Azure reads at a newsreader's pace. The reference videos sit near 2.3 words
 * per second; +12% moves the voice there. Past roughly +20% the consonants
 * start clipping, which is audible as a lisp.
 */
export const DEFAULT_RATE = '+12%';

/**
 * Shape the narration arc instead of reading every line at one level.
 *
 * A neural voice given no direction reads all fourteen lines identically, and
 * fourteen identical lines is what "boring" means — the reference narration is
 * not louder, it moves. The opening leans in, the middle settles into a steady
 * documentary read, and the closing line drops in pitch and slows down, which
 * is the move every documentary narrator makes on a last line.
 */
export function shapeDelivery(index, total) {
  const position = total <= 1 ? 0 : index / (total - 1);
  if (index === 0) return { rate: '+16%', pitch: '+8Hz' };
  if (index === total - 1) return { rate: '+4%', pitch: '-10Hz' };
  if (position > 0.78) return { rate: '+8%', pitch: '-5Hz' };
  // A small alternation through the body keeps consecutive lines from landing
  // on exactly the same contour, which is what reads as recitation.
  return index % 2 ? { rate: '+13%', pitch: '+2Hz' } : { rate: '+11%', pitch: '-2Hz' };
}

export class EdgeTtsError extends Error {
  constructor(message, code) {
    super(message);
    this.name = 'EdgeTtsError';
    this.code = code;
  }
}

function python() {
  return process.env.PYTHON_BIN || (process.platform === 'win32' ? 'python' : 'python3');
}

/**
 * Synthesise every segment in one helper run.
 *
 * Batching matters: each run pays a WebSocket handshake, and Azure throttles a
 * client that reconnects per line — which it answers with a truncated take
 * rather than an error.
 */
export function synthesizeAll({ segments, voice = DEFAULT_VOICE, rate = DEFAULT_RATE, onProgress }) {
  return new Promise((resolve, reject) => {
    const child = spawn(python(), [HELPER], { stdio: ['pipe', 'pipe', 'pipe'] });

    let stdout = '';
    let stderr = '';
    child.stdout.on('data', (d) => {
      stdout += d;
    });
    child.stderr.on('data', (d) => {
      stderr += d;
      for (const line of String(d).split('\n')) {
        const m = line.match(/^dubbing (\d+)\/(\d+)$/);
        if (m) onProgress?.({ stage: 'dubbing', done: Number(m[1]), total: Number(m[2]) });
      }
    });

    child.on('error', (error) => {
      // ENOENT here means Python itself is missing, which needs a different
      // message from "the package is missing" — one is a runtime, one is a
      // pip install.
      reject(
        error.code === 'ENOENT'
          ? new EdgeTtsError(
              `Python 3 tidak ditemukan (${python()}). Pasang Python 3 lalu jalankan ulang.`,
              'python-hilang',
            )
          : error,
      );
    });

    child.on('close', (code) => {
      if (!stdout.trim()) {
        reject(
          new EdgeTtsError(
            `Proses dubbing berhenti (kode ${code}). ${stderr.trim().split('\n').slice(-3).join(' ')}`,
            'proses-gagal',
          ),
        );
        return;
      }
      let payload;
      try {
        payload = JSON.parse(stdout.trim().split('\n').pop());
      } catch {
        reject(new EdgeTtsError(`Balasan dubbing tidak terbaca: ${stdout.slice(0, 200)}`, 'balasan-rusak'));
        return;
      }
      if (!payload.ok) {
        reject(new EdgeTtsError(payload.message || 'Dubbing gagal', payload.error));
        return;
      }
      log.info(`${payload.segments.length} segmen didubbing`, { voice: payload.voice });
      resolve(payload.segments);
    });

    child.stdin.end(
      JSON.stringify({
        voice,
        rate,
        segments: segments.map((s, i) => ({ ...shapeDelivery(i, segments.length), ...s })),
      }),
    );
  });
}

/** Whether this machine can dub right now, and if not, precisely why. */
export async function probe() {
  try {
    await synthesizeAll({
      segments: [],
      onProgress: null,
    });
    return { ready: true };
  } catch (error) {
    return { ready: false, code: error.code, why: error.message };
  }
}

File: src/tts/edge.py
"""
Indonesian dubbing through Microsoft Edge's neural voices.

Invoked by src/pipeline/voice.mjs, not by hand. Reads one JSON request on
stdin, writes the audio files, prints one JSON response on stdout. Progress
goes to stderr so it can be streamed into the UI without corrupting the
response.

Why a Python helper in a Node project: the Edge voice protocol is a WebSocket
handshake with a signed token and a binary frame format, and `edge-tts` is the
maintained implementation of it. Reimplementing that in Node to avoid one
subprocess would mean owning a protocol we did not design and cannot test
against. The subprocess boundary is the cheaper of the two.

Request:
  {"voice": "id-ID-GadisNeural", "rate": "+12%",
   "segments": [{"index": 0, "text": "...", "file": "/abs/path/000.mp3"}]}

Response:
  {"ok": true, "segments": [{"index": 0, "file": "...", "seconds": 2.87,
                             "words": [{"text": "Tas", "start": 0.1, "duration": 0.2}]}]}
"""

import asyncio
import json
import pathlib
import sys

try:
    import edge_tts
except ImportError:
    print(
        json.dumps(
            {
                "ok": False,
                "error": "modul-hilang",
                "message": "Paket edge-tts belum terpasang. Jalankan: pip install edge-tts",
            }
        )
    )
    sys.exit(0)


async def satu(voice: str, rate: str, segment: dict) -> dict:
    """Synthesise one segment and report exactly which words were spoken."""
    # Word-level boundaries are the point. Azure defaults to sentence level,
    # which is too coarse to cut a caption on.
    comm = edge_tts.Communicate(
        segment["text"],
        voice,
        rate=segment.get("rate", rate),
        pitch=segment.get("pitch", "+0Hz"),
        boundary="WordBoundary",
    )

    audio = bytearray()
    words = []
    async for chunk in comm.stream():
        if chunk["type"] == "audio":
            audio.extend(chunk["data"])
        elif chunk["type"] == "WordBoundary":
            # Azure counts in 100-nanosecond ticks; the pipeline plans in seconds.
            words.append(
                {
                    "text": chunk["text"],
                    "start": chunk["offset"] / 1e7,
                    "duration": chunk["duration"] / 1e7,
                }
            )

    if not audio:
        raise RuntimeError("tidak ada audio yang diterima")

    path = pathlib.Path(segment["file"])
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(bytes(audio))

    # The last word boundary, not the container header: a throttled take comes
    # back short but still reports a plausible header duration, and trusting it
    # would let a truncated line through.
    spoken = words[-1]["start"] + words[-1]["duration"] if words else 0.0
    return {
        "index": segment["index"],
        "file": str(path),
        "seconds": round(spoken, 3),
        "words": words,
    }


async def main() -> None:
    request = json.load(sys.stdin)
    voice = request.get("voice", "id-ID-GadisNeural")
    rate = request.get("rate", "+12%")
    segments = request["segments"]

    hasil = []
    # Sequential on purpose: Azure throttles bursts from one client, and a
    # throttled take is returned truncated rather than refused. A slower loop is
    # the difference between a clean take and a silently clipped one.
    for s in segments:
        hasil.append(await satu(voice, rate, s))
        print(f"dubbing {s['index'] + 1}/{len(segments)}", file=sys.stderr, flush=True)

    print(json.dumps({"ok": True, "voice": voice, "rate": rate, "segments": hasil}))


try:
    asyncio.run(main())
except Exception as exc:  # noqa: BLE001 - the message is the product here
    print(json.dumps({"ok": False, "error": "gagal", "message": str(exc)}))
